File: search2dmatrix2.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        #either start at topright or bottomleft corner
        #we have end two conditions as value should oppose on both ways
        #ex. from 15 , check value smaller than 15 (target=5), so traverse previous column to reduce instaed of going to next row as it performs reverse function of increasing value, which is wrong.
        #likewise choose either column or row traversal and reach particular value
        #O(m+n), need to traverse either sides
        #O(1) in place
        
        
        
        #topright corner
        i=0
        j=len(matrix[0])-1
        if matrix:
            #while boundary conditions fulfilled
            while i<=len(matrix)-1 and j>=0:
                #if target is less than value, go to previous column
                if target<matrix[i][j]:
                    j-=1
                #found, >>return
                elif target==matrix[i][j]:
                    return True
                #if target is more than value, go to next row
                else:
                    i+=1
            return False
            

File: .idea/Search2DMatrix2.java
public class Search2DMatrix2 {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0)
            return false;
        int m = matrix.length - 1;
        int n = matrix[0].length - 1;
        int i = 0;
        int j = n;
        while (i <= m && j >= 0) {
            if (matrix[i][j] == target)
                return true;
            else if (matrix[i][j] > target) {
                j--;
            } else
                i++;
        }
        return false;
    }
    public static void main(String args[])
    {
        int k = 45;
        int[][] matrix = {{1,4,7,11,15},{2,5,8,12,19},{3,6,9,16,22},{10,13,14,17,24},{18,21,23,26,30}};
        Search2DMatrix2 ob = new Search2DMatrix2();
        System.out.println(ob.searchMatrix(matrix,k));
    }
}
//time complexity is O(m+n)
//space complexity is O(1)